src/state_machine.zig

```diff
@@ -915,34 +915,60 @@ pub fn StateMachineType(comptime Storage: type, comptime constants_: struct {
             assert(dr_mut.timestamp == dr_immut.timestamp);
             assert(cr_mut.timestamp == cr_immut.timestamp);
 
+            const amount = amount: {
+                var amount = t.amount;
+                if (t.flags.balancing_debit) {
+                    const dr_balance = dr_mut.debits_posted + dr_mut.debits_pending;
+                    const dr_limit = if (dr_immut.flags.debits_must_not_exceed_credits)
+                        dr_mut.credits_posted
+                    else
+                        std.math.maxInt(u64);
+                    amount = std.math.min(amount, dr_limit - dr_balance);
+                }
+
+                if (t.flags.balancing_credit) {
+                    const cr_balance = cr_mut.credits_posted + cr_mut.credits_pending;
+                    const cr_limit = if (cr_immut.flags.credits_must_not_exceed_debits)
+                        cr_mut.debits_posted
+                    else
+                        std.math.maxInt(u64);
+                    amount = std.math.min(amount, cr_limit - cr_balance);
+                }
+
+                // If amount was 0, we will need to return an overflow/exceeds error.
+                break :amount std.math.max(amount, 1);
```

Comment:
Since this logic applies regardless of whether the `balancing` flags are set, it could inadvertently change a `0` amount to a `1`, for a non-balancing transfer.

I think it would be better to have each `balancing` branch break explicitly, eliminating the temporary `amount` variable in the outer scope.

Also, do we check that a transfer cannot be both a `balancing_debit` and a `balancing_credit`?

Reply:
Non-balancing `amount=0` has already been handled in the `amount_must_not_be_zero` branch.

`balancing_debit` and `balancing_credit` are compatible with one another, so we intentionally don't check that.
This is also why the separate balancing branches don't break explicitly.

Reply:
> Non-balancing amount=0 has already been handled in the amount_must_not_be_zero branch.

Yes, but I would rather that we don't depend on that. Lower bar for correctness.

> `balancing_debit` and `balancing_credit` are compatible with one another, so we intentionally don't check that.
This is also why the separate balancing branches don't break explicitly.

I didn't expect that we would support this. So a transfer can both balance the `dr` account's credit balance and at the same time balance the `cr` account's debit balance? Or could you give me an example of where the user would set both flags and what we would do?

Reply:
> So a transfer can both balance the `dr` account's credit balance and at the same time balance the `cr` account's debit balance? Or could you give me an example of where the user would set both flags and what we would do?

Sure! My thinking:

- Setting `balancing_debit` says "transfer as much as possible, up to `amount`, without violating the debit account's limits. That is, a `balancing_debit` transfer will never return `exceeds_debits`, `overflows_debits_pending`, or `overflows_debits`, _unless_ absolutely no money could be transferred.

- Setting `balancing_credit` says "transfer as much as possible, up to `amount`, without violating the _credit_ account's limits .That is, a `balancing_credit` transfer will never return `exceeds_credits`, `overflows_credits_pending`, or `overflows_credits`, _unless_ absolutely no money could be transferred.

- Setting both balancing flags provides the union of these features — transfer as much as possible, up to `amount`, without violating either account's limits. The flags are completely orthogonal — supporting the combination requires no additional code.


Closing an account is probably the most common use case for balancing flags. Closing an account requires only a single flag (balancing debit _or_ credit, depending on which account is being closed), and an amount of `0`.

But balancing transfers are more general than just closing accounts.
For other use cases, there might be account limits on both sides of the transfer.
I don't have a specific application example use case in mind for this right now, but since we support transferring money between accounts where both accounts have limits, I think that we should correspondingly support the combined balancing flags to saturate those same limits.

Reply:
Thanks for a great explanation. Would make excellent detail for the docs.

> The flags are completely orthogonal — supporting the combination requires no additional code.

Beautiful.